apps/namadillo/src/services/extensionEvents/provider.tsx

```diff
@@ -0,0 +1,82 @@
+import { useEventListenerOnce } from "@namada/hooks";
+import { Namada, useIntegration } from "@namada/integrations";
+import { Events, KeplrEvents, MetamaskEvents } from "@namada/types";
+import { useAtomValue, useSetAtom } from "jotai";
+import { createContext } from "react";
+import {
+  addAccountsAtom,
+  balancesAtom,
+  fetchDefaultAccountAtom,
+} from "slices/accounts";
+import { namadaExtensionConnectedAtom } from "slices/settings";
+import { NamadaConnectionRevokedHandler } from "./handlers";
+
+export const ExtensionEventsContext = createContext({});
+
+export const ExtensionEventsProvider: React.FC = (props): JSX.Element => {
+  const namadaIntegration = useIntegration("namada");
+  const keplrIntegration = useIntegration("cosmos");
+  const metamaskIntegration = useIntegration("ethereum");
+  const fetchDefaultAccount = useSetAtom(fetchDefaultAccountAtom);
+  const balances = useAtomValue(balancesAtom);
+  const addAccounts = useSetAtom(addAccountsAtom);
+  const setNamadaExtensionConnected = useSetAtom(namadaExtensionConnectedAtom);
+
+  // Instantiate handlers:
+  const namadaConnectionRevokedHandler = NamadaConnectionRevokedHandler(
+    namadaIntegration as Namada,
+    setNamadaExtensionConnected
+  );
+
+  // Register handlers:
+  useEventListenerOnce(Events.AccountChanged, async () => {
+    await fetchDefaultAccount();
+    balances.refetch();
+  });
+
+  useEventListenerOnce(
+    Events.ConnectionRevoked,
+    namadaConnectionRevokedHandler
+  );
+
+  useEventListenerOnce(KeplrEvents.AccountChanged, async () => {
+    const accounts = await keplrIntegration.accounts();
+    if (accounts) {
+      addAccounts(accounts);
```

Comment:
Just to double-check the intention is to keep all Namada/Keplr/Metamask accounts in the same atom.